Add Spartan ECDSA Rust circuits

0xisk opened the pull request:
# Description

This PR adds Spartan ECDSA prover rust circuits, that uses `wasm-pack` for generating the wasm package.

# Checklist

- [x] I have performed a self-review of my code.
- [x] All existing CI checks pass
  - My code compiles
  - My code follows the style guidelines of this project (code is linted and formatted)
  - My code doesn't break existing tests
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I added a reviewer for this pull request.
- [x] I have assigned myself to this pull request.
- [ ] I have added the appropriate label(s) to this pull request.

Commit:
feat: implementing a wasm-pack liberary for merkle tree

Commit:
faet: implmenting a spartan circuits version that has wasm-pack integrations, and removing halo2 empty project

0xisk committed:
fix: package.json main and types relative paths

0xisk committed:
fix: spartan circuits cargo dependencies

0xisk committed:
fix: remove the built wasm pkg

0xisk committed:
fix: remove deprecated rust crate

0xisk commented:
@sripwoud rebase is done.

0xisk committed:
fix: pnpm lock file

0xisk committed:
fix: package.json workspace and script

0xisk committed:
fix: pnpm lock file

sripwoud commented on circuits/halo2/.gitignore:
Why did you want to delete the whole `circuits/halo2` dir again?
I forgot we had a merkle tree implementation for halo2 there.
How does it compare to what you wrote with poseidon and ark?

Unless it is proven you wrote something faster that makes the halo2 merkle tree implementation redundant, let's keep it in the repo.

0xisk replied:
The `merkle.rs` in the `examples/` dir is not a complete Merkle-tree liberary to be integrated separably with spartan circuit. Also its -as far as I understood- built like a circuit example for Halo2 system which is totally different from Spartan system, the same applied to the Poseidon-gate crate that is used which is built to work with Halo2 and it's source in cargo.toml is https://github.com/axiom-crypto/halo2-lib. 
Finally, this `merkle.rs` doesn't have any of the needed conversions or the expected output that works perfectly with Spartan. Therefore, in our case that is not needed unless we will explore Halo2 again.

sripwoud replied:
@0xisk ok then please push the current `main` branch to a new branch `halo2` to archive it/save it so that we can find it back later in case we need it.
Then it is fine to delete it in this PR and merge it into main

0xisk replied:
https://github.com/anonklub/anonklub/tree/archive/halo2
Done.

sripwoud commented on package.json:
```diff
@@ -34,7 +34,7 @@
   },
   "scripts": {
-    "compile-wasm": "turbo run compile",
+    "compile-wasm": "pnpm --filter @anonklub/merkle-tree compile",
```
I liked it better before.
Why do you want to skip turbo?

`turbo run compile` wouldn't indeed work if called directly in your terminal if you installed it only locally. In a pnpm project, to run the binaries normally in `node_modules/.bin` you need to execute them in a pnpm scripts or use `pnpm exec`.

turbo will pick the scripts called `[name]` in the underlying packages. And then you can wrap `turbo run [name]` in an `pnpm` script.

This executes fine the `@anonklub/spartan-ecdsa` and `@anonklub/merkle-tree` for me
 
```suggestion
    "compile": "turbo run compile"
```

You need to adjust the turbo config.json so that the cache works properly though

0xisk replied:
I see, I agree that this looks better but the only drawback is that you will need to have turbo installed globally. And that is not a good practice, its okay for short-term but for long-term we need a better way I think. 
That's why thought of separating the Rust wasm compilation from the turbo stuff so that I can easily do `pnpm install`.

sripwoud replied:
> you will need to have turbo installed globally

You don't. I don't have turbo installed globally and this script is working fine for me

sripwoud commented on circuits/spartan/src/eth_membership.rs:
Address linting warning `cargo clippy`

sripwoud replied:
And format

sripwoud commented on circuits/spartan/src/eth_membership.rs:
```diff
@@ -0,0 +1,188 @@
+    let pub_key_y_bits = to_le_bits(pub_key.y);
+
+    // We need this transformation because the bits should be in little endian
+    // and the bytes should be in big endian.
```
nit (I had to read this 3 times to see there was a difference :rofl: )
```suggestion
    // We need this transformation because the BITS should be in LITTLE endian
    // and the BYTES should be in BIG endian.
```

sripwoud commented on circuits/spartan/src/utils.rs:
Address lint warnings (`cargo clippy`)

sripwoud commented on circuits/spartan/src/lib.rs:
Address lint warnings and format.

sripwoud commented on circuits/spartan/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+}
+
+#[wasm_bindgen]
+pub fn prove_membership(
```
nit: Given the different casing rules in rust in javascript, we will have some clashes between `cargo clippy` and `eslint`.
It would be nice to either allow non snake case in the rust files, or add later eslint rule exceptions for the TS code. Up to you

0xisk replied:
I will take care of that in a later PR.

sripwoud commented on circuits/spartan/src/utils.rs:
```diff
@@ -0,0 +1,159 @@
+
+// ####################################
+// Helper functions
+// ####################################
```
nit: useless imo (file is called `utils.rs`)
```suggestion
```

sripwoud commented on circuits/spartan/package.json:
```diff
@@ -0,0 +1,22 @@
+        "ts-node": "^10.9.1",
+        "typescript": "5.2.2"
+    }
+}
```
nit: format

prettier doesn't detect it because you have to update the inputs array of the corresponding script in turbo config.json

Note the difference between root script and pkg script
https://turbo.build/repo/docs/core-concepts/monorepos/running-tasks#running-tasks-from-the-root

0xisk replied:
The reason is that `package.json` files are ignored in the `.prettierignore` should we remove it?

sripwoud replied:
yeah I noticed it yesterday. Sorry I should have told you.
I still want to exclude package.json files of the packages that are to be published from format check because `changeset` will edit them (change the version) upon release and messes with existing prettier format.
For consistency then I decided to exclude all package.json from the check.
Let's keep it that way.

Ignore my comment.
I still do like not see the github warning about newline lol. So I do this manually atm.

The right solution would be to define a [prettier override](https://prettier.io/docs/en/configuration.html#configuration-overrides) for the end of line rule.
But let's do this later

sripwoud replied:
capture: https://github.com/anonklub/anonklub/issues/297

sripwoud commented:
And rebase please

Commit:
fix: cargo clippy fixes

Commit:
chore: better comments

sripwoud committed:
fix: build wasm package before install with `devPreinstall` hook (#296)

* fix: build wasm package before install with `devPreinstall` hook

* chore: format

0xisk committed:
fix: adding a build wasm step in the test workflow

0xisk committed:
fix: try to not use compile via turbo in the test workflow

0xisk committed:
fix: adding install wasm-pack step

0xisk committed:
fix: update build rust creates in static analysis

0xisk committed:
chore: lint test workflow

0xisk committed:
fix: make sure to ignore @anonklub/spartan-ecdsa

sripwoud commented on package.json:
```diff
@@ -34,8 +34,9 @@
   "scripts": {
-    "compile-wasm": "turbo run compile",
-    "build": "pnpm run compile-wasm && turbo run build",
+    "pnpm:devPreinstall": "pnpm --filter @anonklub/merkle-tree compile && pnpm --filter @anonklub/spartan-ecdsa compile",
```
~~let's improve with~~
```suggestion
    "pnpm:devPreinstall": "pnpm --filter @anonklub/merkle-tree --filter @anonklub/spartan-ecdsa --parallel compile",
```

sripwoud replied:
bad idea
after testing this breaks the CI action because of temporary file lock: we must not parallelize.
Somehow i didn;t have this issue locally so I that it was a good way to speed up the action

sripwoud commented on .github/workflows/test.yml:
```diff
@@ -61,6 +61,14 @@ jobs:
+
+      - name: Build Rust wasm deps
+        run: pnpm --filter @anonklub/merkle-tree compile && pnpm --filter @anonklub/spartan-ecdsa compile
+
```
~~I think the `devPreinstall` hook makes this step now redundant.~~
```suggestion
```

sripwoud replied:
You were right.
After testing we need to manually run this otherwise the cache is hit. Because it only cares about node deps (pnpm lock file) and knows nothing about the changes that impact rust files...

sripwoud commented on .github/workflows/static-analysis.yml:
```diff
@@ -73,9 +73,8 @@ jobs:
-      - name: Build merkle-tree-wasm deps
-        run: pnpm --filter @anonklub/merkle-tree compile
+      - name: Build Rust wasm deps
+        run: pnpm --filter @anonklub/merkle-tree compile && pnpm --filter @anonklub/spartan-ecdsa compile
```
Same as for test workflow

~~I think the `devPreinstall` hook makes this step now redundant.~~
```suggestion
```

sripwoud replied:
You were right.
After testing we need to manually run this otherwise the cache is hit. Because it only cares about node deps (pnpm lock file) and knows nothing about the changes that impact rust files...

sripwoud committed:
refactor (#298)

* parallelize wasm compile in preInstall hook

* refactor

* to revert

* refactor: do not use jetli action

Install wasm-pack with script instead

Getting some errors otherwise
https://github.com/anonklub/anonklub/actions/runs/7666559819/job/20894718947#step:10:24

* Revert "to revert"

This reverts commit 9b71b5f228ca947ead93810209717a8bf5dceb47.

* fix: do not parallelize compilation of wasm package

running into temp file lock issues otherwise
